### ISSUE TITLE
FXBinder should probably trigger updates inside the UI thread

### ISSUE THREAD

**Issue by aalmiray** [canoo/dolphin-platform#77]
`FxBinder` establishes a soft binding (via listeners) between a DP property and a JavaFX property. Most of the times the JavaFX property belongs to an UI widget. It would simplify matters for a developer if all updates flowing from a DP property to a JavaFX property happened inside the UI thread.

Currently the updates happen inside the same thread that updated the DP property, which is is the UI thread for now, but that's a different design issue one level down enforced by Open Dolphin.

Perhaps `FXBinder` couldhave a setting that lets the developer define if they want the update to happen inside the UI thread or not, to keep behavior compatibility.

**Comment by hendrikebbers** [comment]
I don't know if we need a setting. Why not always check if we are in the Platform thread and otherwise do a invokeAndWait

**Comment by aalmiray** [comment]
Because JavaFX does not have an equivalent to `invokeAndWait`. Well it does, but it's behind a private API that we shouldn't use :smirk:
What we need is `Platform.runLater` if not already inside the UI thread.

**Comment by hendrikebbers** [comment]
I don’t know if runLater is enough. Let’s say a property is set in a background thread. By doing so JFX property will be updated by invokeLater. But maybe the JFX property is bound bidirectional to more than 1 property. By doing so you can end in strange behavior. Example:

dpProperty1;
dpProperty2;
jfxProperty;

FXBinder.bind(jfxProperty).bidirectinalTo(dpProeprty1);
FXBinder.bind(jfxProperty).bidirectinalTo(dpProeprty1);

In external thread call:

dpProperty1.set(„Hello“);
dpProperty2.set(„World“);
System.out.print(dpProperty2);

By doing so the first line will call an update of the Property by using InvokeLater. Than the second one will be executed and again the update will be triggered by using invokeLater. Let’s say the first invokeLater is handled between line 2 and 3. By doing so line 3 will print „Hello“ instead of „World“. That’s why I would like to use a invokeAndWait. Creating this method for JFX is quite simple.
2 years ago I asked Richard Bair why invokeAndWait is not part of the public API. He told me that people that don’t know how to handle it will use it in a wrong / bad way but if you know how to handle it you can simply create it on your own. I think we are part of the second group and therefore should use it ;)

**Comment by netopyr** [comment]
I am afraid, I consider myself in the first group. ;)

Setting a property can cause change listeners to be fired. You have no idea and control over what happens in these listeners, thus wrapping property.set() in an invokeAndWait() seems like a bad idea.

**Comment by hendrikebbers** [comment]
@netopyr can you give a short example what could go wrong? Currently I see no problem and want to understand it

**Comment by netopyr** [comment]
If I understand your idea correctly, you want to wrap a call of the setter of a JavaFX property in an invokeAndWait-block, right? It is a JavaFX property, which means a user can attach any kind of change-listener to it. I could for example create a dead-lock by doing something that waits for the thread that called the setter in the first place.

As a rule of thumb I would never block anything, if I do not have full control over what is going to happen during the block and even if I have full control, I would try to avoid it.

**Comment by aalmiray** [comment]
No,no. This should be done with `invokeLater`semantics, waiting would be overkill. Also, it only makes sense for now to protect the flow of data between DP property and JavaFX property, not the other way around, even in a bidirectional binding scenario.

**Comment by hendrikebbers** [comment]
Ok :)

Von meinem iPhone gesendet

**Comment by hendrikebbers** [comment]
Then this should be quite easy since we already have the UIThreadHandler instance in the config.

Von meinem iPhone gesendet

**Comment by hendrikebbers** [comment]
So the final plan is to call an invokeLater (by using UIThreadHandler) for setting the JFX property based on a change in the DP property, right?

**Comment by aalmiray** [comment]
Yes, although I'd prefer a smarter version of `invokeLater`:
- if already running inside UI thread -> execute
- else post event using `runLater`

**Comment by netopyr** [comment]
+1
The binding should behave synchronously if we are already on the right thread, because it makes it much easier to understand the simple case.

**Comment by hendrikebbers** [comment]
ok
